0011-chunking-o1js.md

```diff
@@ -0,0 +1,95 @@
+
+## Drawbacks
+
+Proving larger circuits will increase proving time and memory consumption. Given the memory limit in most browsers and WASM, this could cause problems for users trying to prove very large circuits.
+
+## Rationale and alternatives
+
+Not exposing Chunking and continuing to rely on recursion for large computations. This adds a significant overhead for developers and adds complexity of applications.
+
+## Prior art
+
+Chunking working in [OCaml and Pickles](https://github.com/MinaProtocol/mina/blob/develop/src/lib/pickles/test/chunked_circuits/test_chunked_circuits.ml) directly.
+
+## Unresolved questions
+
+While trying to prototype Chunking in o1js, I came up with the following unresolved questions. The tests can be found in [this](https://github.com/MinaProtocol/mina/tree/chunking_rfc_unresolved/src/lib/pickles/test/chunked_circuits) branch.
+
+1 Changing the loop that constructs the circuit from `2^17` (roughly `2^16` constraints) iterations to `2^18` iterations (or `2^17` constraints) and also adjusting `num_chunks: 4` and `override_wrap_domain:N1` compiles the program correctly, generates a proof _but_ doesn't verify the proof:
+
+```sh
+Uncaught exception:
+
+  (Pickles.verify dlog_check)
+
+Raised at Base__Error.raise in file "src/error.ml" (inlined), line 8, characters 14-30
+Called from Base__Or_error.ok_exn in file "src/or_error.ml", line 84, characters 17-32
+Called from Dune__exe__Invalid_proof in file "src/lib/pickles/test/chunked_circuits/invalid_proof.ml", line 92, characters 2-9
+```
+
+This can be reproduced by running `dune exec ./invalid_proof.exe` inside of the test folder `src/lib/pickles/test/chunked_circuits`.
+
+Is this behaviors expected? If so, why and how can we prevent the construction of invalid circuits and proofs in o1js? Changing `num_chunks` and `override_wrap_domain` results in an (expected) compilation error - so the chosen values seem to be logical and correct.
+
+On the other hand, setting `num_chunks:8`,`override_wrap_domain:N2` and the loop iteration to `2^19` compiles the circuit and yields a valid proof. (see `valid_proof.ml`)
+
+2 What is the maximum allowed number chunks? From my understanding, setting iteration to `2^20` and `num_chunks:16` requires that the wrap domain must be `16`, but it only accepts `N0 | N1 | N2`. Setting the wrap domain to anything `<= N2` yields an error. How do we deal with large number chunks and constraints? `override_wrap_domain` is of type `Pickles_base.Proofs_verified.t` which only accepts `N0 | N1 | N2`.
+See `large_wrap_domain.ml` for this case. `This circuit was compiled for proofs using the wrap domain of size 15, but the actual wrap domain size for the circuit has size 16`
+
+3 Running a small MVP of Chunking in o1js through the bindings yields the following error: `there are not enough random rows to achieve zero-knowledge (expected: 3, got: 2)`. It seems like some values aren't passed correctly. This also needs to be resolved during implementation.
```

Comment:
Might this be related to the hack/workaround present in the ocaml chunking examples, which is not done in `pickles_bindings.ml`?

```ocaml
(* We must now appease the permutation argument gods, to ensure
   that the 7th permuted column has polynomial degree larger
   than 2^16, and thus that its high chunks are non-zero.
   Suckiness of linearization strikes again!
*)
```

Reply:
It might, although I did try different configurations with different circuits. If this is something that pops up often, we would need a way to better predict when it happens and a way to communicate that to the user if we cant fix it automatically

Reply:
note: I did write the circuit in o1js directly and passed the values manually to `Pickles.compile`, but good point: I will try and run the example via the bindings directly from OCaml to see if the issue persists

Reply:
I meant that maybe we just need to add that workaround to `pickles_bindings.ml`

Reply:
I don't understand what it does, but AFAICT it adds a generic gate with a special kind of witness layout which is different from anything that would be added in a normal circuit

Reply:
yea, but I also wanna make sure that the circuits I test via the bindings are the same as in the OCaml test, so I'll run that same circuit via the OCaml bindings just to check if the problem is fixed by the workaround or if it's something deeper in the bindings

Reply:
Ok. Maybe we're missing proper ZK rows handling in the wasm bindings